File: fib/fib.rs
use std::env;

fn main() {
    let args: Vec<String> = env::args().collect();
    println!("{}", fib2(args[1].parse().unwrap()));
}

/* recursive solution */
fn fib(n: u64) -> u64 {
    match n {
        0 | 1 => n,
        _     => fib(n-1) + fib(n-2),
    }
}

/* tail recursive solution */
fn fib2(n: u64) -> u64 {
    fibonacci2(n, 0, 1)
}

fn fibonacci2(n: u64, a: u64, b: u64) -> u64 {
    match n {
        0 => a,
        1 => b,
        _ => fibonacci2(n-1, b, a+b),
    }
}

File: fib/fib.c
#include <stdio.h>

int fib(int n) {
  if (n < 2) {
    return n;
  } else {
    return fib(n-1) + fib(n-2);
  }
}

int fib2(int n) {
  return n < 2 ? n : fib2(n-1) + fib2(n-2);
}

int main() {
    printf("%d\n", fib2(38));
    return 0;
}
